source/h2_decoder.c

```diff
@@ -923,6 +979,146 @@ static int s_state_fn_frame_unknown(struct aws_h2_decoder *decoder, struct aws_b
+    struct aws_header_block_in_progress *current_block = &decoder->header_block_in_progress;
+    if (current_block->malformed) {
+        goto malformed;
+    }
+
+    if (header_field->name.len == 0) {
+        goto malformed;
+    }
+
+    enum aws_http_header_name name_enum = aws_http_lowercase_str_to_header_name(header_field->name);
+
+    bool is_pseudoheader = header_field->name.ptr[0] == ':';
+    if (is_pseudoheader) {
+        if (current_block->pseudoheaders_done) {
+            DECODER_LOG(ERROR, decoder, "Pseudo-headers must appear before regular fields.");
+            goto malformed;
+        }
+
+        enum pseudoheader_name pseudoheader_enum = s_header_to_pseudoheader_name(name_enum);
+        if (pseudoheader_enum == PSEUDOHEADER_UNKNOWN) {
+            DECODER_LOG(ERROR, decoder, "Unrecognized pseudo-header");
+            goto malformed;
+        }
+
+        /* Ensure request pseudo-headers vs response pseudoheaders were sent appropriately.
+         * This also ensures that request and response pseudoheaders aren't being mixed. */
+        bool expect_request_pseudoheader = decoder->is_server || current_block->is_push_promise;
+        bool is_request_pseudoheader = pseudoheader_enum != PSEUDOHEADER_STATUS;
+        if (expect_request_pseudoheader != is_request_pseudoheader) {
+            DECODER_LOG(ERROR, decoder, "Received wrong type of pseudo-header");
+            goto malformed;
+        }
+
+        /* Protect against duplicates. */
+        if (current_block->pseudoheader_received[pseudoheader_enum]) {
+            DECODER_LOG(ERROR, decoder, "Pseudoheader occurred multiple times");
+            goto malformed;
+        }
+        current_block->pseudoheader_received[pseudoheader_enum] = true;
+
+        /* NOTE: We validate pseudo-headers collectively, after they all arrive, in s_flush_pseudoheaders */
+
+    } else { /* Else regular header-field. */
+
+        /* Regular header-fields come after pseudo-headers, so make sure pseudo-headers are flushed */
+        if (!current_block->pseudoheaders_done) {
+            if (s_flush_pseudoheaders(decoder)) {
+                return AWS_OP_ERR;
+            }
+
+            /* might have realized that header-block is malformed during flush */
+            if (current_block->malformed) {
+                goto malformed;
+            }
+        }
+
+        /* Validate header name (not necessary if string already matched against a known enum) */
+        if (name_enum == AWS_HTTP_HEADER_UNKNOWN) {
+            if (!aws_strutil_is_lowercase_http_token(header_field->name)) {
+                DECODER_LOG(ERROR, decoder, "Header name contains invalid characters");
+                goto malformed;
+            }
+        }
+
+        /* #TODO Validate characters used in header_field->value */
+    }
+
+    /* #TODO should we hold back pseudo-headers until s_flush_pseudoheaders() is called?
+     * - PRO: will never deliver pseudo-header that we later determine is invalid.
```

Comment:
Buffer them. They're not that big, and it makes downstream so much more insane if the headers can be revoked after the fact.